Spatial Video does not show Immersive or View Spatial button when muted or lacks audio track

jyavenard opened the pull request:
#### b062855c649fe25410061d45e175ede7554f760b
<pre>
Spatial Video does not show Immersive or View Spatial button when muted or lacks audio track
<a href="https://bugs.webkit.org/show_bug.cgi?id=288706">https://bugs.webkit.org/show_bug.cgi?id=288706</a>
<a href="https://rdar.apple.com/137636134">rdar://137636134</a>

Reviewed by Jer Noble.

When and how to display Docking, Immersive and View Spatial used the same
heuristic as PiP when in element fullscreen. That is, it always used
the NowPlaying eligible media element which requires that:
- the element has an audible audio track
- is not muted
- is currently playing or has been played previously.

This NowPlaying eligible media would be used when docking even if it wasn&apos;t
in the hierarchy of the element in fullscreen.
As such, we couldn&apos;t dock or play in immersive or spatial mode silent videos.

We now decouple the handling of element fullscreen from this heuristic.
Instead we will only consider all video elements that are a child of the element
in fullscreen.
To avoid having the UI process regularly query the web process, we now maintain
in each of the VideoPresentationModels the fullscreen inheritance state and keep
those in sync between the webcontent and UI processes.

Should more than one video element be visible while in element fullscreen
for now we will select the one that is the most visible in the current viewport
and fill over 25% of the web view area (similar to the heuristic adopted in 280710@main).
Whenever the fullscreen visibility change, we will update the status
of the top icons and display them accordingly.

This heuristic will be further tuned in a future change (such as preferring
a video that was the most recently played).
For now, this new heuristic is placed behind a pref which is off by default.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/DocumentFullscreen.cpp:
(WebCore::hasJSEventListener): Added.
(WebCore::DocumentFullscreen::dispatchPendingEvents): Only check for JS listeners to determine which even to fire.
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
(WebCore::VideoPresentationModelClient::isChildOfElementFullscreenChanged):
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.h:
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
(WebCore::VideoPresentationModelVideoElement::cleanVideoListeners): Fly-by: Keep separate weakRef to Document
so that we properly unregister listeners if HTMLMediaElement was GCed.
(WebCore::VideoPresentationModelVideoElement::setVideoElement):
(WebCore::VideoPresentationModelVideoElement::updateForEventName):
(WebCore::VideoPresentationModelVideoElement::documentFullscreenChanged):
(WebCore::VideoPresentationModelVideoElement::documentObservedEventNames):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
(WebCore::VideoPresentationInterfaceIOS::playbackSessionModel const):
(WebCore::VideoPresentationInterfaceIOS::setSpatialImmersive):
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
(VideoFullscreenControllerContext::isChildOfElementFullscreen const):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _videosInElementFullscreenChanged]):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::setDocumentVisibility):
(WebKit::VideoPresentationManagerProxy::setIsChildOfElementFullscreen):
(WebKit::VideoPresentationManagerProxy::videosInElementFullscreenChanged):
(WebKit::VideoPresentationManagerProxy::createModelAndInterface):
(WebKit::VideoPresentationManagerProxy::ensureModelAndInterface):
(WebKit::VideoPresentationManagerProxy::ensureModel):
(WebKit::VideoPresentationManagerProxy::ensureInterface):
(WebKit::VideoPresentationManagerProxy::findModelAndInterface const):
(WebKit::VideoPresentationManagerProxy::findInterface const):
(WebKit::VideoPresentationManagerProxy::returningToStandbyInterface const):
(WebKit::VideoPresentationManagerProxy::setHasVideo):
(WebKit::VideoPresentationManagerProxy::setVideoDimensions):
(WebKit::VideoPresentationManagerProxy::enterFullscreen):
(WebKit::VideoPresentationManagerProxy::bestVideoForElementFullscreen):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::videosInElementFullscreenChanged):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::videosInElementFullscreenChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::videosInElementFullscreenChanged):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController initWithWebView:]):
(-[WKFullScreenViewController videoControlsManagerDidChange]):
(-[WKFullScreenViewController videosInElementFullscreenChanged]):
(-[WKFullScreenViewController _bestVideoPresentationInterface]):
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):
(-[WKFullScreenViewController viewDidAppear:]):
(-[WKFullScreenViewController viewWillTransitionToSize:withTransitionCoordinator:]):
(-[WKFullScreenViewController _enterVideoFullscreenAction:]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController videosInElementFullscreenChanged]):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationInterfaceContext::isChildOfElementFullscreenChanged):
(WebKit::VideoPresentationManager::isChildOfElementFullscreenChanged):

Canonical link: <a href="https://commits.webkit.org/291482@main">https://commits.webkit.org/291482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b91d65858a3adaab59a59cfe5155b3ba98df719f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93041 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12592 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2286 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98040 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43567 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95091 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12873 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21046 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71138 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28558 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96043 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84187 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51467 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9398 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1821 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42880 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/85751 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79689 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1798 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100065 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91707 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20093 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80161 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20345 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79463 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19728 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24012 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1316 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13168 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20077 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25253 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114356 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19764 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33004 "Found 12 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.mini-mode, microbenchmarks/memcpy-wasm-small.js.bytecode-cache, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-slow-memory, wasm.yaml/wasm/stress/repro_1289.js.default-wasm, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager-jettison, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate-catch.js.default-wasm, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.default-wasm ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23224 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21505 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
```diff
@@ -166,6 +170,33 @@
+        return;
+
+    for (auto& client : copyToVector(m_clients))
+        client->documentFullscreenChanged(m_isChildOfElementFullscreen);
```
I think we should call this client method `isChildOfElementFullscreenChanged()`

jernoble commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.messages.in:
```diff
@@ -29,6 +29,7 @@
 messages -> VideoPresentationManagerProxy {
     SetHasVideo(WebKit::PlaybackSessionContextIdentifier contextId, bool hasVideo)
     SetDocumentVisibility(WebKit::PlaybackSessionContextIdentifier contextId, bool isDocumentVisible)
+    SetDocumentFullscreen(WebKit::PlaybackSessionContextIdentifier contextId, bool isChildOfElementFullscreen)
```
Ditto, but `SetIsChildOfElementFullscreen(...)`.

jernoble commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -261,8 +261,26 @@ - (BOOL)prefersStatusBarHidden
+    if (RefPtr interface = findInterface(contextId)) {
         interface->documentVisibilityChanged(isDocumentVisible);
+        Ref model = ensureModel(contextId);
+        if (model->isChildOfElementFullscreen())
+            videosInElementFullscreenChanged();
```
Why are we updating `videosInElementFullscreenChanged()` when the document becomes visible or not?

jyavenard replied:
I understood this method as being the video element becoming visible or not. As such it impacts the calculation of the most visible video element.

jernoble commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -965,6 +983,9 @@ - (BOOL)prefersStatusBarHidden
     }
+
+    if (ensureModel(secondContextId)->isChildOfElementFullscreen())
+        videosInElementFullscreenChanged();
```
I don't think this is necessary; this method is about native fullscreen, not Element fullscreen, and swapping native fullscreen states won't effect whether the video element has a fullscreen ancestor.

jernoble commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1044,14 +1065,17 @@ - (BOOL)prefersStatusBarHidden
+    if (auto interface = findInterface(contextId)) {
         interface->hasVideoChanged(hasVideo);
+        ensureModel(contextId)->m_hasVideo = hasVideo;
+    }
```
We should probably add something like `findInterfaceAndModel(...)` so functions like this don't have to search the hash map twice.

jernoble commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1060,6 +1084,8 @@ - (BOOL)prefersStatusBarHidden
         m_mockPictureInPictureWindowSize.setHeight(DefaultMockPictureInPictureWindowWidth / videoDimensions.aspectRatio());
         return;
     }
+    if (videosInElementFullscrenChanged)
+        this->videosInElementFullscreenChanged();
```
Maybe we need a different name here. The videos in element fullscreen haven't changed, but we do need to re-evaluate the list of videos. Maybe `updateVideosInElementFullscreen()`?

jyavenard replied:
It is really to say that one of the videos in the element fullscreen changed. So we need to recheck which one is the most visible. 

I guess I could store the context Id of the last video we made primary and only recheck if that video changed

jernoble reviewed (approved):
r=me with nits

jyavenard committed:
Spatial Video does not show Immersive or View Spatial button when muted or lacks audio track
https://bugs.webkit.org/show_bug.cgi?id=288706
rdar://137636134

Reviewed by Jer Noble.

When and how to display Docking, Immersive and View Spatial used the same
heuristic as PiP when in element fullscreen. That is, it always used
the NowPlaying eligible media element which requires that:
- the element has an audible audio track
- is not muted
- is currently playing or has been played previously.

This NowPlaying eligible media would be used when docking even if it wasn't
in the hierarchy of the element in fullscreen.
As such, we couldn't dock or play in immersive or spatial mode silent videos.

We now decouple the handling of element fullscreen from this heuristic.
Instead we will only consider all video elements that are a child of the element
in fullscreen.
To avoid having the UI process regularly query the web process, we now maintain
in each of the VideoPresentationModels the fullscreen inheritance state and keep
those in sync between the webcontent and UI processes.

Should more than one video element be visible while in element fullscreen
for now we will select the one that is the most visible in the current viewport
and fill over 25% of the web view area (similar to the heuristic adopted in 280710@main).
Whenever the fullscreen visibility change, we will update the status
of the top icons and display them accordingly.

This heuristic will be further tuned in a future change (such as preferring
a video that was the most recently played).
For now, this new heuristic is placed behind a pref which is off by default.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/DocumentFullscreen.cpp:
(WebCore::hasJSEventListener): Added.
(WebCore::DocumentFullscreen::dispatchPendingEvents): Only check for JS listeners to determine which even to fire.
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
(WebCore::VideoPresentationModelClient::isChildOfElementFullscreenChanged):
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.h:
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
(WebCore::VideoPresentationModelVideoElement::cleanVideoListeners): Fly-by: Keep separate weakRef to Document
so that we properly unregister listeners if HTMLMediaElement was GCed.
(WebCore::VideoPresentationModelVideoElement::setVideoElement):
(WebCore::VideoPresentationModelVideoElement::updateForEventName):
(WebCore::VideoPresentationModelVideoElement::documentFullscreenChanged):
(WebCore::VideoPresentationModelVideoElement::documentObservedEventNames):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
(WebCore::VideoPresentationInterfaceIOS::playbackSessionModel const):
(WebCore::VideoPresentationInterfaceIOS::setSpatialImmersive):
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
(VideoFullscreenControllerContext::isChildOfElementFullscreen const):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _videosInElementFullscreenChanged]):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::setDocumentVisibility):
(WebKit::VideoPresentationManagerProxy::setIsChildOfElementFullscreen):
(WebKit::VideoPresentationManagerProxy::videosInElementFullscreenChanged):
(WebKit::VideoPresentationManagerProxy::createModelAndInterface):
(WebKit::VideoPresentationManagerProxy::ensureModelAndInterface):
(WebKit::VideoPresentationManagerProxy::ensureModel):
(WebKit::VideoPresentationManagerProxy::ensureInterface):
(WebKit::VideoPresentationManagerProxy::findModelAndInterface const):
(WebKit::VideoPresentationManagerProxy::findInterface const):
(WebKit::VideoPresentationManagerProxy::returningToStandbyInterface const):
(WebKit::VideoPresentationManagerProxy::setHasVideo):
(WebKit::VideoPresentationManagerProxy::setVideoDimensions):
(WebKit::VideoPresentationManagerProxy::enterFullscreen):
(WebKit::VideoPresentationManagerProxy::bestVideoForElementFullscreen):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::videosInElementFullscreenChanged):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::videosInElementFullscreenChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::videosInElementFullscreenChanged):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController initWithWebView:]):
(-[WKFullScreenViewController videoControlsManagerDidChange]):
(-[WKFullScreenViewController videosInElementFullscreenChanged]):
(-[WKFullScreenViewController _bestVideoPresentationInterface]):
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):
(-[WKFullScreenViewController viewDidAppear:]):
(-[WKFullScreenViewController viewWillTransitionToSize:withTransitionCoordinator:]):
(-[WKFullScreenViewController _enterVideoFullscreenAction:]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController videosInElementFullscreenChanged]):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationInterfaceContext::isChildOfElementFullscreenChanged):
(WebKit::VideoPresentationManager::isChildOfElementFullscreenChanged):

Canonical link: https://commits.webkit.org/291482@main
